File: 0536-construct-binary-tree-from-string/0536-construct-binary-tree-from-string.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* str2tree(string s) {
        if(s.length() == 0) return nullptr;
        
        int i = 0, n = s.length();
        
        while(i < n && s[i] != '(')
        {
            i++;
        }
        
        // cout << s.substr(0, i) << endl;
        
        if(i == n) return new TreeNode(stoi(s));
        TreeNode* root = new TreeNode(stoi(s.substr(0, i)));
        
        int balance = 0;
        int start = i;
        
        int j = i;
        
        while(j < n)
        {
            char c = s[j];
            if(c == '(') balance++;
            else if(c == ')') balance--;
            
            if(balance == 0)
            {
                if(start == i)
                {
                    root->left = str2tree(s.substr(start + 1, j - start - 1));
                    start = j + 1;
                }
                else
                {
                    root->right = str2tree(s.substr(start + 1, j - start - 1));
                }
            }
            
            j++;
        }
        
        return root;
    }
};

File: 0536-construct-binary-tree-from-string/0536-construct-binary-tree-from-string.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def str2tree(self, s: str) -> Optional[TreeNode]:
        if not s:
            return None
        i = 0
        n = len(s)
        
        while i < n and s[i] != '(':
            i += 1
        
        if i == n:
            return TreeNode(int(s))
        
        node = TreeNode(int(s[:i]))
        
        j = i
        balance = 0
        start = i
        
        while j < n:
            if s[j] == '(':
                balance += 1
            elif s[j] == ')':
                balance -= 1
            
            
            if balance == 0 and start == i:
                node.left = self.str2tree(s[start + 1:j])
                start = j + 1
                
            elif balance == 0:
                node.right = self.str2tree(s[start + 1:j])
            j += 1
        return node
            
        